Added examples of larger workunit size processing

DTL2020 opened the pull request:
With SSE2 and AVX2 V-resizers, SSE2 8bit V-resize with double sized of workunit shows about +12% more performance in 1 threaded test and AVX2 float32 V-resize of quad-sized of workunit shows about +70% performance (over 1x in 3.7.4 release).

The H-resizers are somehow more complex to increase of workunit size at least to double size and the expected performance boot may be about same +15..20% as we have with V-resizers (integer 8 bit processing). 

Currently most performance boost we got with V-float32 AVX2 resize looks like left completely not optimized in 3.7.4 release,

DTL2020 commented:
Test script for testing V float32 AVX2 resize performance used:
BlankClip(10000, 100, 1000, color=$B0B0B0)

ConvertToPlanarRGB()
ConvertBits(32)

LanczosResize(width, height*40)

qyot27 commented:
Purely cosmetic changes (whitespace, line formatting, etc.) need to be in their own commit so the actual changes can be readily seen.

pinterf commented:
Since the resizers are still being under heavy construction on my machine - I'm gonna wait with your changes, nevertheless, I will test them later and integrate the ideas if needed.

DTL2020 commented:
This may be not considered as final and best pull-request - this only some very limited 'tech demo' of the main and simple general idea to all (very many versions of SSE/AVX) resizing functions: The 'workunit size' to be processed in the each loop spin is critical enough to performance and its increasing may give more or less visible performance benefit (depending on how much non-optimized was initial version of the function). But the total performance benefit based on the workunit size is non-linear depend on its value (also expected to have some visible maximum - but it can be different on different execution CPU chips). And increasing this value make program text larger and somehow more complex (while it mostly copy of existing lines). Unfortunately C-compilers are not smart enough yet to make this adjustments at the compilation time (it may look somehow close to 'loop unroll' operation - we decrease total number of loop spins at software runtime as we increase data size processed in each loop spin). But it is up to the final developer decision if it worth to spend time on the increasing size of program text while receiving some performance benefit (at some tests). Like +1% performance with 2x size of function text may be 'not worth'. While +10% with 2..3x size of function test may be 'still worth' and +50% may be 'definitely worth'. Sadly looks we still do not have enough AI-driven compilers to to all this work at compile time and test-driven optimization processing. So currently it is task for program writer.

DTL2020 commented:
I tried to mark it as 'draft' as it mostly some very few working examples of suggested changes. Also hope they not add new bugs and issues (still very few testing performed).

DTL2020 commented:
Also in some of SIMD functions I found unusual uncached store instruction like 
https://github.com/AviSynth/AviSynthPlus/blob/2e34f5daba7f223e8172e28ed7900b0997749207/avs_core/filters/intel/resample_avx2.cpp#L519C2-L519C19

It may flush data from all CPU caches into host RAM and if data is required quickly in the next processing (next filter ?) it may be slow to read back. 

Also some documentation states some 'fence' (store fence) may be needed later but it is not visible in current program text:
https://doc.rust-lang.org/core/arch/x86_64/fn._mm_stream_si128.html
Safety of non-temporal stores

After using this intrinsic, but before any other access to the memory that this intrinsic mutates, a call to [_mm_sfence](https://doc.rust-lang.org/core/arch/x86/fn._mm_sfence.html) must be performed by the thread that used the intrinsic. In particular, functions that call this intrinsic should generally call _mm_sfence before they return.

Usage of cached or not cached stores from the output of each filter in a chain is a subject of complex testing in real AVS use cases (with a set of more or less complex scripts) to show if it good or not. Simple AVSmeter test with skipping of output data may give too nice results but in a sequence of filters it may cause bad memory subsystem performance. May be this instruction simply come from some copy-paste from other software with different use cases.

pinterf commented:
I don't know, sometimes using stream is giving quicker benchmark, sometimes not. I remember such questions in mvtools2 as well. It depended on too many things. In one scenario it was better, in another one just the opposite. One processor arhitecture favored this, a newer one not. Ideally, we should recompile an Avisynth and plugins for different scripts. One for qtgmc, one for frequently used Expr, etc. As for one benchmark we use Intel or LLVM, and for some others pure MSVC (but I admit, the latter case is rare :) ) 

After having seen and benchmarking the result of different compilers and processor architectures, I'd say that no good solution exist. 

E.g. this is a fresh adventure with LLVM: I put in only a single one extra line in the horizontal resizer (16 bit case needs to be shifted from unsigned to signed 16 bit range, so adding -32768 is needed). Everybody knows and admits that this is more complex, compared to the 10 bit case. However, inserting this extra step modified somehow the optimizer's flow (register use and allocation or whatever) so that it produced almost 3x (!) quicker code than the 10 bit case, the - theoretically - simpler one. Of course, I did not want to accept this illogical issue. I have rewritten the code, reordered the lines, since I wanted to understand it, why it is done, Compiled it with gcc, msvc, checked even the asm output, run through godbolt compiler explorer and finally turned out nothing. LLVM had a very smart and lucky decision for the more complex code.

DTL2020 commented:
The most typical error in testing performance is testing 'raw' performance
of isolated functions with sort of null input and null output. It only
tests RAW computing performance and not full performance in a real chain of
processing between lots of chained functions.

As I see in my tests of new versions of V-float resize with increased 'SIMD
workunit size' - it saturates by something (most likely memory subsystem
performance) with about 2..3 cores/threads from I7(5?)-9600K CPU. So most
RAW compute benefit is visible only in single thread and single core
testing.

It also means the optimization of memory subsystem performance of AVS
filters is very important in any big multithreading environment at the very
poor performance of current end-users hosts (or at least unbalanced
performance of computing cores and memory subsystems).

So for AVS filters testing with typical frame sizes about CPU caches size
for 1..3 threads it is important to develop some close to real use cases
test workflow with 'real source' and 'real sink' filters to see if used
memory read/write model do not cause total performance lost when using
filter in real environment. Maybe even develop some debug-developing
typical source+sink filters to set before and after filters to test. Though
it also can significantly depend on AVS caching in between filters in a
graph.

For several-stages RAM read-write filters like the current dual-1D AVS core
resampling engine for 2D buffers it may be also better to use
'patch/block-based' processing instead of a full-frame processing of big
enough buffers sizes when we can easily have workflow of
1. Read buffer from host RAM
2. Process H or V and store it in memory. Because it's too large to store,
it finally goes to the slowest host RAM.
3. Read immediate H or V processed from host RAM
4. Process second dimension with H or V (or even rotate 90 degrees to use a
single processing engine and rotate back with also host RAM storage).
5. Store to memory (if use stream-uncached store - it also hints to store
to host RAM).
6. Next filter - read from host RAM (also slow)...

For possibly better processing of big enough buffers in 'patch-block'
method:
1. Read part of buffer to cache (L1/L2/L3).
2. Process H or V and store to L1/L2 cache.
3. Read immediate H or V processed from cache (much faster and do not touch
slow host RAM bus).
4. Process second dimension (or make 90 degrees to and from rotation) with
cached data.
5. Finally store to host RAM

Method 2 only uses 2 host RAM full buffer transfers and may be faster. The
size of the patch/block to process is subject of testing (may be close to
L1D cache size of about 20..40 KBytes or L2/L3 cache of several MBytes).
Also it is good to make patch size configurable by user or auto-optimized
from cache size and threads count so all threads can use total CPU cache
without overflow and lost performance).

Method 2 also requires making the resampler more complex and to solve
possible edge issues at the patches/blocks borders. Maybe with some
overlapping processing and cutting out only the good internal part of the
processed buffer able to be seamlessly merged into the output frame. The
'bad' part of the buffer is expected to have size about filter_support/2 or
even up to full size of filter_support. So the size of the patch-block to
process must be significantly more in comparison with filter_support too
(also may be an argument to the auto-selection of the patch-block size
function). Possibly this can be partially script-simulated with the new
filter MultiOverlay.

On Wed, Apr 23, 2025 at 1:52 PM Ferenc Pintér ***@***.***>
wrote:

> I don't know, sometimes using stream is giving quicker benchmark,
> sometimes not. I remember such questions in mvtools2 as well. It depended
> on too many things. In one scenario it was better, in another one just the
> opposite. One processor arhitecture favored this, a newer one not. Ideally,
> we should recompile an Avisynth and plugins for different scripts. One for
> qtgmc, one for frequently used Expr, etc. As for one benchmark we use Intel
> or LLVM, and for some others pure MSVC (but I admit, the latter case is
> rare :) )
>
> After having seen and benchmarking the result of different compilers and
> processor architectures, I'd say that no good solution exist.
>
> E.g. this is a fresh adventure with LLVM: I put in only a single one extra
> line in the horizontal resizer (16 bit case needs to be shifted from
> unsigned to signed 16 bit range, so adding -32768 is needed). Everybody
> knows and admits that this is more complex, compared to the 10 bit case.
> However, inserting this extra step modified somehow the optimizer's flow
> (register use and allocation or whatever) so that it produced almost 3x (!)
> quicker code than the 10 bit case, the - theoretically - simpler one. Of
> course, I did not want to accept this illogical issue. I have rewritten the
> code, reordered the lines, since I wanted to understand it, why it is done,
> Compiled it with gcc, msvc, checked even the asm output, run through
> godbolt compiler explorer and finally turned out nothing. LLVM had a very
> smart and lucky decision for the more complex code.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/AviSynth/AviSynthPlus/pull/435#issuecomment-2823887740>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQMGLMZKWGH6W6D7R5TMNIT225WFZAVCNFSM6AAAAAB3OS7LH2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMRTHA4DONZUGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
> *pinterf* left a comment (AviSynth/AviSynthPlus#435)
> <https://github.com/AviSynth/AviSynthPlus/pull/435#issuecomment-2823887740>
>
> I don't know, sometimes using stream is giving quicker benchmark,
> sometimes not. I remember such questions in mvtools2 as well. It depended
> on too many things. In one scenario it was better, in another one just the
> opposite. One processor arhitecture favored this, a newer one not. Ideally,
> we should recompile an Avisynth and plugins for different scripts. One for
> qtgmc, one for frequently used Expr, etc. As for one benchmark we use Intel
> or LLVM, and for some others pure MSVC (but I admit, the latter case is
> rare :) )
>
> After having seen and benchmarking the result of different compilers and
> processor architectures, I'd say that no good solution exist.
>
> E.g. this is a fresh adventure with LLVM: I put in only a single one extra
> line in the horizontal resizer (16 bit case needs to be shifted from
> unsigned to signed 16 bit range, so adding -32768 is needed). Everybody
> knows and admits that this is more complex, compared to the 10 bit case.
> However, inserting this extra step modified somehow the optimizer's flow
> (register use and allocation or whatever) so that it produced almost 3x (!)
> quicker code than the 10 bit case, the - theoretically - simpler one. Of
> course, I did not want to accept this illogical issue. I have rewritten the
> code, reordered the lines, since I wanted to understand it, why it is done,
> Compiled it with gcc, msvc, checked even the asm output, run through
> godbolt compiler explorer and finally turned out nothing. LLVM had a very
> smart and lucky decision for the more complex code.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/AviSynth/AviSynthPlus/pull/435#issuecomment-2823887740>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQMGLMZKWGH6W6D7R5TMNIT225WFZAVCNFSM6AAAAAB3OS7LH2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMRTHA4DONZUGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

pinterf commented:
I'm gonna return to this and your post later, now I commited my work, and I'm fed up with benchmarks and the differently behaving compilers and mainly with MSVC but also with gcc. So I just copy-paste here what I've written. One can make tests infinitely, and you'll always find a case there one will decrease in speed, the other gets better, much better, then you are satisfied, make another test on another machine which ruins all your former ideas. Then you realize that though you spent weeks on it, you cannot teach a dog to sing to enroll in the Opera (=msvc making quick code from C in the integer domain). 

Giga-benchmarks. Built, modified the code and measured at least 1000 times. (o.k. for build count the figures are somewhat less)

Benchmarks.

```
  SetMaxCPU("none")
  #SetMaxCPU("SSSE3")
  #SetMaxCPU("AVX2")
  ColorbarsHD(200,200, pixel_type="YUV444P8") # P8, P10, P16, PS

  wmul=2 # >1: horizontal resizer test
  hmul=1 # >1: vertical resizer test

  avs=LanczosResize(width*wmul, height*hmul, taps=16)
  avsr=z_ConvertFormat(width=width*wmul, height=height*hmul, resample_filter="lanczos", filter_param_a=16)
  fmtconv=fmtc_resample(w=width*wmul, h=height*hmul, kernel="lanczos", taps=16)
  avs
```

Figures in general has no place in source code but it is useful to have a clue 
and see the huge differences between compilers and code variants.
Also, only mazochists want to test integer optimization with MSVC (x86). Wasted time.
Use llvm on Arm, clangcl/llvm on Windows x86.

```
Horizontals                Hor+Vert    Verticals
 [8]  [10-14]  [16]  [32]  [8]  [10]   [8]  [10-14] [16] [32]
 177     170    101  208               262   486    417   217    C-RaspberryPi5 gcc 12.2 (code variant)
 137     469     88  208                                         C-RaspberryPi5 gcc 12.2 (code variant)
 227     218    147  178                                         C-RaspberryPi5 gcc 12.2 no vector attrib ??! Quicker than vector attrib version gcc not recommended
 416     611    579  404    128  186   362   609    556   624    C-RaspberryPi5 llvm 14 vector attrib
  90     185     94  403                                         C-RaspberryPi5 llvm 14 vector attrib + integer madd@H
 180     182    162          67        213   212    206   639    C-RaspberryPi5 llvm 14 no vector attrib
1230    1168   1129                                              C-ClangCl SSE2
1270    1238   1186                   1550  1555   1560  3670    C-ClangCl AVX2
1051    1126   1102                                              C-Intel ICX 2025 SSE2
1513    2355   1560 1128                                 1969    C-Intel ICX 2025 SSE4.2  smart madd!
2173    2620   1804 1283                                         C-Intel ICX 2025 AVX2    3.7.6 old tmp
1938    2413   1775 1061    442  453  1136  1126   1037  3511    C-Intel ICX 2025 AVX2
 212     188    187  264     73   64   223   195    198   268    C-MSVC SSE2     3.7.3 code
 417     463    360                    449   424    384   352    C-MSVC SSE2     3.7.4 code
 215     215     97  928     79   79   220   744     96  1951    C-MSVC SSE2 
 201     206     99                                              C-MSVC AVX2 
 597     631    651                                              C-Intel SSE4.2  3.7.4 code
1183    1193    889                                              C-Intel AVX2    3.7.4 code
5600                       2140                                  SIMD-avsresize (AVX2 or AVX512?)
2260                        840                                  SIMD-fmtconv (16 bit output for 8 bits) 
4578    2614   2560 2250   1490       4220  4534   3887  3570    SIMD-MSVC AVX2 3.7.3 (horizontal was memory-boundary unsafe)
3631    3505   3221 2344   1291 1354  3804  4466   3855  2260    SIMD-MSVC AVX2 3.7.4 Float vertical regression - no time to finish
3720    3478   3130 2385   1566 1480  5014  5288   5077  3810    SIMD-MSVC AVX2 + incrementing offsets in V, 20-25% gain in integer verticals
4730    4612   4233 2487   1390 1471  3792  4476   4380  3942    SIMD-ClangCl AVX2, verticals behind MSVC by surprise
2373    2181   1893 1306    868  723  2660  2137   2670  1886    SIMD-MSVC SSSE3 + incrementing offsets in V, 20-25% gain in integer verticals
2294    2979   2595 1460    859  976  2623  2865   2625  1962    SIMD-Intel ICX 2025 SSSE3
4395    4616   4160 2570   1664 1720  5110  5870   5085  2999    SIMD-Intel ICX 2025 AVX2 Surprisingly slow at vertical float FIXME, slower than C :)
```

* float has different optimization: 8 pixels 2 coeffs
** on aarch64 the float benchmarks included a ConvertBits(8) at the end.

Non-x86 platforms are the focus of this comparison, as they lack SIMD optimized paths.
x86 platforms have their own SIMD optimized paths and do not use C.

Compilers can give totally different results even after some reordering of the code,
See gcc aarch64: the 8 bit case dropped to 2/3 speed while 16 bits increased by a factor of 250%
depending on where I put a line.

Considerations for aarch64 (Armv8a, includes neon, e.g. RPi5 with gcc 12.2, llvm 14.0.6)
- gcc is not recommended, their vectorizer is either broken or not yet ready.
  Using vector attribute gave slower code.
- llvm gave consistent fast results.
- Vertical resampler: aarch64 + GCC, (8-16 bits) the 4-pixel 4-coefficient version performs best.

On x86 platforms MSVC is unusable regarding integer vectorization, the more the code helps to
recognize vectorization patterns the slower assembly it produces, e.g. horizontal resizer AVX2 16 bit: 
MSVC 99 fps, MSVC+clangcl: 1186 fps, Intel+LLVM: 1775 fps. Joke.

MSVC on x86 is only capable of vectorizing 32 bit float code, and even then it is not the fastest.
There is no reason to not use clangcl for x86 Visual Studio builds. It comes officially with VS2022 for free.

DTL2020 commented:
"  wmul=2 # >1: horizontal resizer test
  hmul=1 # >1: vertical resizer test"

I see the input size is 200x200 but what were wmul and hmul at the tests shown ? The sad thing may happen - you got very nice optimized 200x200 to 400x400 resize fit in CPU caches in 1 threaded mode (raw computing optimized) but when it come to 4K or 8K frames processing and/or user sets all-cores (6..8..16 and more) Prefetch(N) - all efforts may be hardly limited by memory subsystem and it may be 'second stage' of the optimizing of large buffers and all-cores threads processing. In the best case the most AI-driven compiler + profile-giuded optimization (available in the Intel compiler only ?) may somehow 'understand' the complexity of the task and attempt to cut total large task to pieces (in the case of H+V resizes) in 'auto' mode instead of doing this manually. I agree the test of the large-frame performance of full H+V resize and all-cores multithreading  may be put to later stages of resampling engine optimization.

pinterf commented:
For testing, I used a width and height multiplier of 2.

Anyway, I think it's extremely fortunate that resampling is a rare and fast operation. This prevents me from spending another two weeks' effort to achieve a marginal speed gain. There are way too many variables to consider: checking all relevant compilers, possibly upcoming architectures, testing for single threading, JPSDR-like multi-threading, and Avisynth-like MT. Additionally, smaller or larger thread counts must be tested, and it must work for SD, HD, and 4K resolutions. When embedded into a larger script, as usual, any effort we put in will likely be nullified and made invisible, either due to caching (or a lack thereof) or because other filters are heavier than the resampling.

Imagine my benchmark table inserted into the former post with 3 or more additional dimensions and more free variables.

So yes, as you mentioned, there must be a completely different algorithm for these various cases. It's a very interesting theme for a university graduate. :) Interesting topic, and never-ending, which is why I could not stop working on it for weeks.

I think the first tech demos in this area must initially work in C and as an external filter, to see the costs, caveats, and feasibility. Once that's done, optimization can begin.

DTL2020 commented:
" I think it's extremely fortunate that resampling is a rare and fast
operation.  "

Resampling is the very widely used operation in current and future planned
updates for sub-sample precision processing. One example I plan to post
some day in the https://github.com/AviSynth/AviSynthPlus/issues/433 feature
request. Currently it only has implementation with shift-resampling. But it
is not perfect because it causes some shift of original content of the
image. But it is a simple tech-demo of the required effect on simple 2
levels transient and fast enough because it uses resampling engine only
once and without size change. Better solution is also based on heavy usage
of resampler engine - Generic super-sampling (interpolation) approach:
1. Upsize (interpolate) image area to process to some order of 10 (the
precision of processing) - like 10x for 0.1 precision or 100x for 0.01
precision and so on.
2. Apply required operation in integer coordinate domain (input float
coordinate parameter multiplied to the interpolation factor 10 or 100 or
more and truncate fractional part).
3. Downsize back to original size (sample down to original spatial
resolution).

This approach is universal for any AVS+ filter requiring any geometric
operation with float-domain coordinates (like
Layer/Overlay/LetterBox/AddBorders and any other with coordinate input).
And maybe any new geometric transform filters like rotation to angle.

So a single resampling engine is an important building block for many
operations with better precision and quality. It is worth putting possible
efforts on optimizing.

"I think the first tech demos in this area must initially work in C and as
an external filter, to see the costs, caveats, and feasibility. Once that's
done, optimization can begin."

Yes - it can be called like Name_ResizeTiled() or Name_ResizeBig() to show
it is designed to process big buffers. Will try to start with the simplest
script-based form to see how it is working and what is the size of the
possibly damaged (not equal to single full-frame resample) area at the
edges of a tile. Even currently implemented workarounds may not work
completely perfect on the original frame cut to pieces  As the external
resize engine may be JPSDR's resize may be changed because it may be too
hard (for me) to extract the resampling engine from AVS+ core.



On Sun, Apr 27, 2025 at 5:52 PM Ferenc Pintér ***@***.***>
wrote:

> *pinterf* left a comment (AviSynth/AviSynthPlus#435)
> <https://github.com/AviSynth/AviSynthPlus/pull/435#issuecomment-2833496317>
>
> For testing, I used a width and height multiplier of 2.
>
> Anyway, I think it's extremely fortunate that resampling is a rare and
> fast operation. This prevents me from spending another two weeks' effort to
> achieve a marginal speed gain. There are way too many variables to
> consider: checking all relevant compilers, possibly upcoming architectures,
> testing for single threading, JPSDR-like multi-threading, and Avisynth-like
> MT. Additionally, smaller or larger thread counts must be tested, and it
> must work for SD, HD, and 4K resolutions. When embedded into a larger
> script, as usual, any effort we put in will likely be nullified and made
> invisible, either due to caching (or a lack thereof) or because other
> filters are heavier than the resampling.
>
> Imagine my benchmark table inserted into the former post with 3 or more
> additional dimensions and more free variables.
>
> So yes, as you mentioned, there must be a completely different algorithm
> for these various cases. It's a very interesting theme for a university
> graduate. :) Interesting topic, and never-ending, which is why I could not
> stop working on it for weeks.
>
> I think the first tech demos in this area must initially work in C and as
> an external filter, to see the costs, caveats, and feasibility. Once that's
> done, optimization can begin.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/AviSynth/AviSynthPlus/pull/435#issuecomment-2833496317>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQMGLM74FTP44OCJ5GXQ4IL23TVKFAVCNFSM6AAAAAB3OS7LH2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMZTGQ4TMMZRG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

DTL2020 commented:
I start some performance test of AVS+ 3.7.5 resampler with AVSmeter and
found some very strange feature:

All resamplers and all frame sizes completely do not benefit from CPU
caches. It also applies to avsresize.

So it looks like some significant design issues still present either in
AVS+ processing logic or in AVSmeter way of performance testing.

Some tests were at old Core2Duo E7500 CPU (pre-AVX) and some at i5-9600K
CPU with AVX2.

Test script is
ColorbarsHD(320,320, pixel_type="YUV444P8")
#ColorbarsHD(3200,3200, pixel_type="YUV444P8")

mul=4

LanczosResize(width*mul, height*mul, taps=1)
LanczosResize(width/mul, height/mul, taps=1)

(about the same with simplest PointResize)

The i5-9600K L2 cache average bandwidth is about 500 GB/s and its external
RAM is about 25 GB/s. So the ratio of the memory subsystem performance is
about 500/25=20. The 320x320 x4 plane fits in 1.5 MB L2 cache and 3200x3200
x4 plane is about 4x larger than the L3 cache (9 MB).

Expected resample results - the 320x320 plane processing is at least 2
times faster when being processed with 2-stages H+V resampler (with
90rotating - even more faster) if measured in samples per second, Because
intermediate H or V resample result must be readback to processing from 20x
faster cache. And the 3200x3200 frame size result must be about 2 times
slower because the intermediate result must go to host RAM and readback via
the slowest bus in the system
All test results - the large frame not fitted in any cache - performance is
even slightly better.
320x320 frame size - 300fps
3200x3200 frame size - 3 fps
Both show equal samples per second processing rate of 320x320 x4 (also x3
planes) x300fps - about 368 Msamples/s.
Same as 3200x3200 x4 (also x3 planes) x3fps.

I run many tests with slow steps, increasing of frame size and different
taps for Lanczos - no one test anytime shows any benefit from much faster
CPU caches for small frame sizes. It looks like very slow and purely
'compute bound' processing but it is unlikely can be so at least for
simplest possible PointResize and for Lanczos(taps=1).

It is a subject of some research why we have such poor CPU caches usage by
AVS+ processing software. It looks like something not allowed to use CPU
caches at all  (and I do not think it is a rare usage of uncached-stores
instructions). After this will be solved - we can continue to make more
memory-optimized processing.

I will try to download the latest sources and look in the debugger how the
resampler data processing logic is designed to attempt to understand why we
have so poor results even inside single 2stages filter processing. I
understand AVS+ cache logic may cause all usable data flushing from CPU
caches into host RAM because of frames processing queues ?. But it should
not affect the single filters running in several stages of the single plane
data processing in a sequence.

On Sun, Apr 27, 2025 at 7:18 PM Andrey Glazunov ***@***.***>
wrote:

> " I think it's extremely fortunate that resampling is a rare and fast
> operation.  "
>
> Resampling is the very widely used operation in current and future planned
> updates for sub-sample precision processing. One example I plan to post
> some day in the https://github.com/AviSynth/AviSynthPlus/issues/433
> feature request. Currently it only has implementation with
> shift-resampling. But it is not perfect because it causes some shift of
> original content of the image. But it is a simple tech-demo of the required
> effect on simple 2 levels transient and fast enough because it uses
> resampling engine only once and without size change. Better solution is
> also based on heavy usage of resampler engine - Generic super-sampling
> (interpolation) approach:
> 1. Upsize (interpolate) image area to process to some order of 10 (the
> precision of processing) - like 10x for 0.1 precision or 100x for 0.01
> precision and so on.
> 2. Apply required operation in integer coordinate domain (input float
> coordinate parameter multiplied to the interpolation factor 10 or 100 or
> more and truncate fractional part).
> 3. Downsize back to original size (sample down to original spatial
> resolution).
>
> This approach is universal for any AVS+ filter requiring any geometric
> operation with float-domain coordinates (like
> Layer/Overlay/LetterBox/AddBorders and any other with coordinate input).
> And maybe any new geometric transform filters like rotation to angle.
>
> So a single resampling engine is an important building block for many
> operations with better precision and quality. It is worth putting possible
> efforts on optimizing.
>
> "I think the first tech demos in this area must initially work in C and as
> an external filter, to see the costs, caveats, and feasibility. Once that's
> done, optimization can begin."
>
> Yes - it can be called like Name_ResizeTiled() or Name_ResizeBig() to show
> it is designed to process big buffers. Will try to start with the simplest
> script-based form to see how it is working and what is the size of the
> possibly damaged (not equal to single full-frame resample) area at the
> edges of a tile. Even currently implemented workarounds may not work
> completely perfect on the original frame cut to pieces  As the external
> resize engine may be JPSDR's resize may be changed because it may be too
> hard (for me) to extract the resampling engine from AVS+ core.
>
>
>
> On Sun, Apr 27, 2025 at 5:52 PM Ferenc Pintér ***@***.***>
> wrote:
>
>> *pinterf* left a comment (AviSynth/AviSynthPlus#435)
>> <https://github.com/AviSynth/AviSynthPlus/pull/435#issuecomment-2833496317>
>>
>> For testing, I used a width and height multiplier of 2.
>>
>> Anyway, I think it's extremely fortunate that resampling is a rare and
>> fast operation. This prevents me from spending another two weeks' effort to
>> achieve a marginal speed gain. There are way too many variables to
>> consider: checking all relevant compilers, possibly upcoming architectures,
>> testing for single threading, JPSDR-like multi-threading, and Avisynth-like
>> MT. Additionally, smaller or larger thread counts must be tested, and it
>> must work for SD, HD, and 4K resolutions. When embedded into a larger
>> script, as usual, any effort we put in will likely be nullified and made
>> invisible, either due to caching (or a lack thereof) or because other
>> filters are heavier than the resampling.
>>
>> Imagine my benchmark table inserted into the former post with 3 or more
>> additional dimensions and more free variables.
>>
>> So yes, as you mentioned, there must be a completely different algorithm
>> for these various cases. It's a very interesting theme for a university
>> graduate. :) Interesting topic, and never-ending, which is why I could not
>> stop working on it for weeks.
>>
>> I think the first tech demos in this area must initially work in C and as
>> an external filter, to see the costs, caveats, and feasibility. Once that's
>> done, optimization can begin.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/AviSynth/AviSynthPlus/pull/435#issuecomment-2833496317>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AQMGLM74FTP44OCJ5GXQ4IL23TVKFAVCNFSM6AAAAAB3OS7LH2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMZTGQ4TMMZRG4>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

pinterf commented:
Yep, 433 is not forgotten.

I experienced the same x10 speed difference, both for a 8 bit 444 and a single plane Y. For single plane I just inserted a simple ExtractY, which is a zero-cost filter in this case (SubFrame logic).

I fed this single threaded avs script into an almost default configured x265 encoding (`x265.exe  --crf 16 --preset ultrafast --output "s1_out.hevc" --input "s5.avs"`) which seemed to be using all my 16 cores on an i7-11700).
It showed the same ~10x difference. Fps values are by avsmeter64 (no final read of the result), followed in parenthesis with the fps shown by x265 encoder.

```
#                       YUV444P8        Y8
# 320x320 mul 4 :      377 (310)     1135 (952fps)
# 3200x3200 mul 4 :      3,65 (3,2)    11,13 (9,87fps)
```

Yes, if your expectations are correct, for small frame sizes we'd see significantly different samples per second values, which is not the observed case. Isn't it because, even in the simplest scenario, the bottleneck is not memory access but processing (due to unresolved instruction dependencies)?

Regarding avsresize, I think it is a compound filter with multiple conversion options, where the filters are likely chained together (resampling, color, bit depth). However, for simple integer resampling, I think it remains in the integer domain; otherwise, the data may be processed in floats. A similar logic applies to fmtconv internals.

DTL2020 commented:
> even in the simplest scenario, the bottleneck is not memory access but processing (due to unresolved instruction dependencies)?

As the next tests show - the most slow part of current dual-D resize is H-resize. And it looks the main point of the compute-bounding of the performance. But because V-resize is about 10x times faster - the limit is not AVX2 unit compute performance but non-optimal design of H-resize for SIMD. I hope it can be changed in new H-resize SIMD designs and after we got about equal performance of V and H resizers - the tests of memory optimizations may be continued. I think the H-resize can be even faster in comparison with V-resize because it uses single read stream of source data.  But currently H-resize SIMD implementation is more complex in comparison with V-resize but still very slow.

pinterf commented:
1.) Idea for V-resizer, large kernel size case, where x, x+stride, s+2*stride .... x + (kernel_size - 1)*stride starts again for the next 16 pixels, there is no chance to reuse the cache again. Instead, we could do meandering, once we reached the last line, we would reverse the direction and for the bunch of pixel processing , we could fetch x + 16 + (kernel_size-1)*stide then x + 16 + (kernel_size-2)*stride, ... x+16 positions. With this, at least the couple of the extreme lines can have chance to read from cache.

I don't remember now what are the usual kernel sizes, so maybe this feature - if works - is just a nice to have, because in real world nobody is using that. But it is interestion enough for us to try out :)

2.) I suppose H has a heavy register pressure, even on AVX2, there is an Intel Compiler optimizer report for that. Ideally the less "spills" (too few registers and if we are shoft of them, compiler stores them temporarily on stack). The less we see of this counter, the quicker code we have. The very first - two pixel processing at a time - resulted in significant gain in performance, I'd keep this double processing option. But the concept of processing 8 pixels at a time (and holding their temporary results) can be decreased to 4 and see what happens. Because this is an unrolled loop.

Note: everything is force inlined. I tested that I kept some of these forceinlined modules in real functions, but it became super slow, possibly parameter passing killed the performance and the register use optimizations. Plus: when everything is inlined, the code and the loop is much larger, so even the code prefetch and caching would also suffer. But I'm not sure with this statement.

DTL2020 commented:
1. For better cache usage I mean using intermediate previous resizer output
as input to the next stage.
Two cases possible (for the example H-resize first and V-resize last):
1.1 H-resize output fits the CPU caches (all levels at least) - we simply
need to start V-resize processing from this plane again (not as currently
we have Y-plane always first and chroma next in every resize stage).
1.2. H-resize output is larger than any CPU cache - switching to tiled
processing with tile size optimized for storage of intermediate (H-resized)
results in CPU cache. This will save the system from storing H-result in
host RAM and readback.

2. For H-resize algorithm: It is the most common resize for any 1D Digital
Signal Processing like audio and any other 1D data processing so I think an
infinite number of implementations exist also for SIMD of today's
mass-usage CPUs (like intel/AMD). So it looks AVS simply got some unlucky
and complex but SIMD-unfriendly implementation and we can start from simply
testing different possible H-resize implementations. Not trying to fix this
complex but slow. As it seen V-resize is much more simple in program text
and much faster and easy enough to fine-tune for register file size and
memory data exchange. So it may be simply poor design idea for currently
available SIMD machines in currently used H-resize in AVS and can be
skipped if other design ideas how to calculate simple convolution for 1D
data linearly stored in RAM will be much faster. As I read all intel SIMD
designs were optimized for DSP tasks and some libraries (free open source)
may already exist.

On Thu, May 1, 2025 at 8:27 AM Ferenc Pintér ***@***.***>
wrote:

> *pinterf* left a comment (AviSynth/AviSynthPlus#435)
> <https://github.com/AviSynth/AviSynthPlus/pull/435#issuecomment-2844134524>
>
> 1.) Idea for V-resizer, large kernel size case, where x, x+stride,
> s+2*stride .... x + (kernel_size - 1)*stride starts again for the next 16
> pixels, there is no chance to reuse the cache again. Instead, we could do
> meandering, once we reached the last line, we would reverse the direction
> and for the bunch of pixel processing , we could fetch x + 16 +
> (kernel_size-1)*stide then x + 16 + (kernel_size-2)*stride, ... x+16
> positions. With this, at least the couple of the extreme lines can have
> chance to read from cache.
>
> I don't remember now what are the usual kernel sizes, so maybe this
> feature - if works - is just a nice to have, because in real world nobody
> is using that. But it is interestion enough for us to try out :)
>
> 2.) I suppose H has a heavy register pressure, even on AVX2, there is an
> Intel Compiler optimizer report for that. Ideally the less "spills" (too
> few registers and if we are shoft of them, compiler stores them temporarily
> on stack). The less we see of this counter, the quicker code we have. The
> very first - two pixel processing at a time - resulted in significant gain
> in performance, I'd keep this double processing option. But the concept of
> processing 8 pixels at a time (and holding their temporary results) can be
> decreased to 4 and see what happens. Because this is an unrolled loop.
>
> Note: everything is force inlined. I tested that I kept some of these
> forceinlined modules in real functions, but it became super slow, possibly
> parameter passing killed the performance and the register use
> optimizations. Plus: when everything is inlined, the code and the loop is
> much larger, so even the code prefetch and caching would also suffer. But
> I'm not sure with this statement.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/AviSynth/AviSynthPlus/pull/435#issuecomment-2844134524>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQMGLM2R3RKVAXN2QRL3BGD24GWE7AVCNFSM6AAAAAB3OS7LH2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQNBUGEZTINJSGQ>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>
